Add google cloud endpoints chart

ultimateboy opened the pull request:
Two things about this chart which might prevent it from being accepted:

1. `Must successfully launch with default values (helm install .)`
What if the chart is useless without a customized value? I notice the minecraft chart requires you to set a variable to accept the EULA, so this doesn't seem like a completely strict requirement. Cloud Endpoints proxies API requests to a backend API. Without information about the location of that API, this chart does not function.

2. `Must include source GitHub repositories for images used in the Chart`
What if the image is provided by Google and there is no known source? I have no idea where the source for the endpoints image is :/

Commit:
Add google cloud endpoints chart

ultimateboy commented:
I signed it!

jackzampolin commented on stable/gcloud-endpoints/templates/deployment.yaml:
```diff
@@ -0,0 +1,83 @@
+          image: "{{ .Values.image }}"
+          imagePullPolicy: {{ default "" .Values.imagePullPolicy | quote }}
+          args: [
+{{- if (.Values.httpPort) }}
```
Can these be set using a Configmap? This looks a little messy and doesn't allow for changes without restarting the nginx process

ultimateboy replied:
Due to the dual use of the values in both args and ports, I feel like this works better. I'm not certain how a configmap would make this any cleaner.

Also, the way the httpport value defaults if the sslport or http2port are both empty seems tricky to accomplish in any other way.

jackzampolin commented on stable/gcloud-endpoints/templates/deployment.yaml:
```diff
@@ -0,0 +1,83 @@
+{{- end }}
+          ]
+          ports:
+{{- if (.Values.httpPort) }}
```
Can you implement `.Values.httpPort` as an array? Then you could support any number of ports:
```
{{ $port, $name := range .Values.httpPort }}
  - containerPort: {{ $port }}
     name: {{ $name }}

```

ultimateboy replied:
But then I wouldn't be able to re-use the ports in the args block.

jackzampolin commented on stable/gcloud-endpoints/templates/service.yaml:
```diff
@@ -0,0 +1,33 @@
+  name: {{ template "fullname" . }}
+spec:
+  ports:
+{{- if (.Values.httpPort) }}
```
See above comment about array of ports.

ultimateboy replied:
See above.

jackzampolin commented on stable/gcloud-endpoints/values.yaml:
```diff
@@ -0,0 +1,75 @@
+## Kubernetes configuration
+## For minikube, set this to NodePort, elsewhere use LoadBalancer
+##
+serviceType: LoadBalancer
```
Switch to ClusterIP by default per @viglesiasce

ultimateboy replied:
Updated

prydonius commented:
> 1. Must successfully launch with default values (helm install .)
> What if the chart is useless without a customized value? I notice the minecraft chart requires you to set a variable to accept the EULA, so this doesn't seem like a completely strict requirement. Cloud Endpoints proxies API requests to a backend API. Without information about the location of that API, this chart does not function.

This is indeed not a requirement we're strictly enforcing currently, and we should consider removing it.

> Must include source GitHub repositories for images used in the Chart
> What if the image is provided by Google and there is no known source? I have no idea where the source for the endpoints image is :/

I'm not sure how strictly we want to enforce this requirement, but for now adding any documentation link as a source would be useful. For example, I found https://cloud.google.com/endpoints/docs/quickstart-compute-engine-docker#running_the_api_and_extensible_service_proxy_in_a_docker_container

viglesiasce commented:
For 1) we are generally having the chart install without deploying resources, then having the NOTES.txt print out instructions on how to set the right values. Once those are set the install proceeds (see gitlab-ce and ghost). 

2) I'm fine with @prydonius suggestion to include that link.

Testing this now.

viglesiasce commented on stable/gcloud-endpoints/templates/deployment.yaml:
```diff
@@ -0,0 +1,83 @@
+            timeoutSeconds: 1
+{{- end }}
+          resources:
+            {{ include "toYaml" .Values.resources | indent 10 }}
```
This indent needs to be 12 otherwise the chart wont install.

viglesiasce commented on stable/gcloud-endpoints/templates/NOTES.txt:
```diff
@@ -0,0 +1,21 @@
+1. Get the Endpoints URL by running:
```
Can you add some more info on where people can go to find out more about Cloud Endpoints. Is there a sample app that you can point them to? Or maybe a tutorial on getting an app setup that they can leverage this chart for?

prydonius commented:
@ultimateboy any updates on this?

lachie83 commented:
@ultimateboy any update?

ultimateboy commented:
I will be addressing the issues brought up shortly. Sorry for the delay.

ultimateboy commented:
I've addressed the above comments. I added multiple links to cloud endpoints docs to both README and NOTES.

Good timing as Cloud Endpoints just went GA today :) https://cloudplatform.googleblog.com/2017/02/Google-Cloud-Endpoints-now-GA-a-fast-scalable-API-gateway.html?m=1

lachie83 commented:
@k8s-bot ok to test

ultimateboy commented:
Just realized the image tag changed from `0.3` to `1`. Updated.
